File: platforms/apple/Sources/NimbusTests/MochaTests.swift
//
// Copyright (c) 2020, Salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo
// root or https://opensource.org/licenses/BSD-3-Clause
//

import JavaScriptCore
import WebKit
import XCTest
@testable import Nimbus

class MochaTests: XCTestCase, WKNavigationDelegate {
    class MochaTestBridge {
        init(webView: WKWebView) {
            self.webView = webView
        }

        let webView: WKWebView
        let expectation = XCTestExpectation(description: "testsCompleted")
        var failures: Int = -1
        func testsCompleted(failures: Int) {
            self.failures = failures
            expectation.fulfill()
        }

        func onTestFail(testTitle: String, errMessage: String) {
            NSLog("[\(testTitle)] failed: \(errMessage)")
        }

        func ready() {}
        func sendMessage(name: String, includeParam: Bool) {
            if includeParam {
                webView.broadcastMessage(name: name, arg: MochaMessage())
            } else {
                webView.broadcastMessage(name: name)
            }
        }
    }

    var webView: WKWebView!

    var loadingExpectation: XCTestExpectation?

    override func setUp() {
        webView = WKWebView()
        webView.navigationDelegate = self
    }

    override func tearDown() {
        webView.navigationDelegate = nil
        webView = nil
    }

    func loadWebViewAndWait() {
        loadingExpectation = expectation(description: "web view loaded")
        loadingExpectation?.assertForOverFulfill = false
        if let path = Bundle(for: MochaTests.self).url(forResource: "nimbus", withExtension: "js", subdirectory: "iife"),
            let nimbus = try? String(contentsOf: path) {
            let userScript = WKUserScript(source: nimbus, injectionTime: .atDocumentStart, forMainFrameOnly: true)
            webView.configuration.userContentController.addUserScript(userScript)
        } else {
            // when running from swiftpm, look for the file relative to the source root
            let basepath = URL(fileURLWithPath: #file)
            let url = URL(fileURLWithPath: "../../../../packages/nimbus-bridge/dist/iife/nimbus.js", relativeTo: basepath)
            if FileManager().fileExists(atPath: url.absoluteURL.path), let script = try? String(contentsOf: url) {
                let userScript = WKUserScript(source: script, injectionTime: .atDocumentStart, forMainFrameOnly: true)
                webView.configuration.userContentController.addUserScript(userScript)
            }
        }

        if let url = Bundle(for: MochaTests.self).url(forResource: "index", withExtension: "html", subdirectory: "test-www") {
            webView.loadFileURL(url, allowingReadAccessTo: url)
        } else {
            // when running from swiftpm, look for the file relative to the source root
            let basepath = URL(fileURLWithPath: #file)
            let url = URL(fileURLWithPath: "../../../../packages/test-www/dist/test-www/index.html", relativeTo: basepath)
            if FileManager().fileExists(atPath: url.absoluteURL.path) {
                webView.loadFileURL(url.absoluteURL, allowingReadAccessTo: url.absoluteURL)
            }
        }
        wait(for: [loadingExpectation!], timeout: 5)
    }

    func webView(_: WKWebView, didFinish _: WKNavigation!) {
        loadingExpectation?.fulfill()
    }

    // This is necessary because WebViewBridge has an optimization
    // to only add a single user script per plugin and this set of tests
    // uses WebViewConnection without WebViewBridge
    func addUserScript(connection: WebViewConnection) {
        if let script = connection.userScript() {
            let userScript = WKUserScript(source: script, injectionTime: .atDocumentStart, forMainFrameOnly: false)
            webView.configuration.userContentController.addUserScript(userScript)
        }
    }

    func testExecuteMochaTests() {
        let testBridge = MochaTestBridge(webView: webView)
        let webBridge = BridgeBuilder.createBridge(for: webView, plugins: [])
        let connection = WebViewConnection(from: webView, bridge: webBridge, as: "mochaTestBridge")
        connection.bind(testBridge.testsCompleted, as: "testsCompleted")
        connection.bind(testBridge.ready, as: "ready")
        connection.bind(testBridge.sendMessage, as: "sendMessage")
        connection.bind(testBridge.onTestFail, as: "onTestFail")
        addUserScript(connection: connection)
        let callbackTestPlugin = CallbackTestPlugin()
        let callbackConnection = WebViewConnection(from: webView, bridge: webBridge, as: callbackTestPlugin.namespace)
        callbackTestPlugin.bind(to: callbackConnection)
        addUserScript(connection: callbackConnection)
        let apiTestPlugin = JSAPITestPlugin()
        let apiTestConnection = WebViewConnection(from: webView, bridge: webBridge, as: apiTestPlugin.namespace)
        apiTestPlugin.bind(to: apiTestConnection)
        addUserScript(connection: apiTestConnection)

        loadWebViewAndWait()

        webView.evaluateJavaScript("""
        const titleFor = x => x.parent ? `${titleFor(x.parent)} ${x.title}` : x.title
        mocha.run(failures => { __nimbus.plugins.mochaTestBridge.testsCompleted(failures); })
             .on('fail', (test, err) => __nimbus.plugins.mochaTestBridge.onTestFail(titleFor(test), err.message));
        true;
        """) { _, error in

            if let error = error {
                XCTFail(error.localizedDescription)
            }
        }

        wait(for: [testBridge.expectation], timeout: 30)
        XCTAssertEqual(testBridge.failures, 0, "Mocha tests failed: \(testBridge.failures)")
    }
}

public class JSContextMochaTests: XCTestCase {
    var context: JSContext = JSContext()

    class JSContextMochaTestBridge {
        let context: JSContext
        let globals: JavaScriptCoreGlobalsProvider
        let expectation = XCTestExpectation(description: "testsCompleted")
        var failures: Int = -1

        init(context: JSContext) {
            self.context = context
            globals = JavaScriptCoreGlobalsProvider(context: context)
            let nimbusDeclaration = """
            __nimbus = {"plugins": {}};
            true;
            """
            context.evaluateScript(nimbusDeclaration)
        }

        func testsCompleted(failures: Int) {
            self.failures = failures
            expectation.fulfill()
        }

        func onTestFail(testTitle: String, errMessage: String) {
            NSLog("[\(testTitle)] failed: \(errMessage)")
        }

        func ready() {}
        func sendMessage(name: String, includeParam: Bool) {
            guard let broadcastFunc = context.globalObject.objectForKeyedSubscript("__nimbus")?.objectForKeyedSubscript("broadcastMessage") else {
                XCTFail("couldn't get a reference to broadcastMessage")
                return
            }
            if includeParam {
                let encodedMessage = try! JSValueEncoder().encode(MochaMessage(), context: context) // swiftlint:disable:this force_try
                broadcastFunc.call(withArguments: [name, encodedMessage])
            } else {
                broadcastFunc.call(withArguments: [name])
            }
        }
    }

    public override func setUp() {
        context = JSContext()
    }

    func evalScript(name: String, ext: String, context: JSContext) {
        if let path = Bundle(for: JSContextMochaTests.self).url(forResource: name, withExtension: ext, subdirectory: "test-www"),
            let script = try? String(contentsOf: path) {
            NSLog("eval-ing \(name).\(ext) from test-www into context")
            context.evaluateScript(script)
        } else {
            let basepath = URL(fileURLWithPath: #file)
            let url = URL(fileURLWithPath: "../../../../packages/test-www/dist/test-www/\(name).\(ext)", relativeTo: basepath)
            if FileManager().fileExists(atPath: url.absoluteURL.path), let script = try? String(contentsOf: url) {
                context.evaluateScript(script)
            }
        }
    }

    func loadContext() {
        // we might need to set something as 'global' before loading chai
        let global = context.globalObject
        context.setObject(global, forKeyedSubscript: "global" as NSString)
        context.evaluateScript("global.location = { \"search\": \"\" };")
        evalScript(name: "mocha", ext: "js", context: context)
        context.evaluateScript("mocha.reporter('json'); mocha.setup('bdd');")
        evalScript(name: "chai", ext: "js", context: context)
        evalScript(name: "bundle", ext: "js", context: context)
    }

    func testExecuteMochaTests() {
        let testBridge = JSContextMochaTestBridge(context: context)
        let contextBridge1 = BridgeBuilder.createBridge(for: JSContext(), plugins: [])
        let connection = JSContextConnection(from: context, bridge: contextBridge1, as: "mochaTestBridge")
        connection.bind(testBridge.testsCompleted, as: "testsCompleted")
        connection.bind(testBridge.ready, as: "ready")
        connection.bind(testBridge.sendMessage, as: "sendMessage")
        connection.bind(testBridge.onTestFail, as: "onTestFail")
        let callbackTestPlugin = CallbackTestPlugin()
        let contextBridge2 = BridgeBuilder.createBridge(for: JSContext(), plugins: [])
        let callbackConnection = JSContextConnection(from: context, bridge: contextBridge2, as: callbackTestPlugin.namespace)
        callbackTestPlugin.bind(to: callbackConnection)
        let apiTestPlugin = JSAPITestPlugin()
        let contextBridge3 = BridgeBuilder.createBridge(for: JSContext(), plugins: [])
        let apiTestConnection = JSContextConnection(from: context, bridge: contextBridge3, as: apiTestPlugin.namespace)
        apiTestPlugin.bind(to: apiTestConnection)

        loadContext()

        let testScript = """
        const titleFor = x => x.parent ? `${titleFor(x.parent)} ${x.title}` : x.title
        mocha.run(failures => { __nimbus.plugins.mochaTestBridge.testsCompleted(failures); })
             .on('fail', (test, err) => __nimbus.plugins.mochaTestBridge.onTestFail(titleFor(test), err.message));
        true;
        """

        let evalResult = context.evaluateScript(testScript)

        wait(for: [testBridge.expectation], timeout: 30)
        XCTAssertNotNil(evalResult, "test script failed to execute")
        XCTAssertEqual(testBridge.failures, 0, "Mocha tests failed: \(testBridge.failures)")
    }
}

struct MochaMessage: Encodable {
    var stringField = "This is a string"
    var intField = 42
}

public struct TestError: Error, Encodable {
    let code: Int
    let message: String
}

public class CallbackTestPlugin {
    func callbackWithSingleParam(completion: @escaping (MochaMessage) -> Swift.Void) {
        let mochaMessage = MochaMessage()
        completion(mochaMessage)
    }

    func callbackWithTwoParams(completion: @escaping (MochaMessage, MochaMessage) -> Swift.Void) {
        var mochaMessage = MochaMessage()
        mochaMessage.intField = 6
        mochaMessage.stringField = "int param is 6"
        completion(MochaMessage(), mochaMessage)
    }

    func callbackWithSinglePrimitiveParam(completion: @escaping (Int) -> Swift.Void) {
        completion(777)
    }

    func callbackWithTwoPrimitiveParams(completion: @escaping (Int, Int) -> Swift.Void) {
        completion(777, 888)
    }

    func callbackWithPrimitiveAndUddtParams(completion: @escaping (Int, MochaMessage) -> Swift.Void) {
        completion(777, MochaMessage())
    }

    func promiseResolved() -> String {
        return "promise"
    }

    func promiseRejectedEncoded() throws -> String {
        throw TestError(code: 42, message: "mock promise rejection")
    }

    func promiseRejected() throws -> String {
        throw MockError.rejectedError
    }
}

enum MockError: Error {
    case rejectedError
}

extension CallbackTestPlugin: Plugin {
    public var namespace: String {
        return "callbackTestPlugin"
    }

    public func bind<C>(to connection: C) where C: Connection {
        connection.bind(callbackWithSingleParam, as: "callbackWithSingleParam")
        connection.bind(callbackWithTwoParams, as: "callbackWithTwoParams")
        connection.bind(callbackWithSinglePrimitiveParam, as: "callbackWithSinglePrimitiveParam")
        connection.bind(callbackWithTwoPrimitiveParams, as: "callbackWithTwoPrimitiveParams")
        connection.bind(callbackWithPrimitiveAndUddtParams, as: "callbackWithPrimitiveAndUddtParams")
        connection.bind(promiseResolved, as: "promiseResolved")
        connection.bind(promiseRejectedEncoded, as: "promiseRejectedEncoded")
        connection.bind(promiseRejected, as: "promiseRejected")
    }
}

class JSAPITestPlugin: Plugin {
    func nullaryResolvingToInt() -> Int {
        return 5
    }

    func nullaryResolvingToIntArray() -> [Int] {
        return [1, 2, 3]
    }

    func nullaryResolvingToObject() -> JSAPITestStruct {
        return JSAPITestStruct()
    }

    func unaryResolvingToVoid(param: Int) {
        XCTAssertEqual(param, 5)
    }

    func unaryObjectResolvingToVoid(param: JSAPITestStruct) {
        XCTAssertEqual(param, JSAPITestStruct())
    }

    func binaryResolvingToIntCallback(param: Int, completion: (Int) -> Void) {
        XCTAssertEqual(param, 5)
        completion(5)
    }

    func binaryResolvingToObjectCallback(param: Int, completion: (JSAPITestStruct) -> Void) {
        XCTAssertEqual(param, 5)
        completion(JSAPITestStruct())
    }

    func binaryResolvingToObjectCallbackToInt(param: Int, completion: (JSAPITestStruct) -> Void) -> Int {
        XCTAssertEqual(param, 5)
        completion(.init())
        return 1
    }

    var namespace: String {
        return "jsapiTestPlugin"
    }

    func bind<C>(to connection: C) where C: Connection {
        connection.bind(nullaryResolvingToInt, as: "nullaryResolvingToInt")
        connection.bind(nullaryResolvingToIntArray, as: "nullaryResolvingToIntArray")
        connection.bind(nullaryResolvingToObject, as: "nullaryResolvingToObject")
        connection.bind(unaryResolvingToVoid, as: "unaryResolvingToVoid")
        connection.bind(unaryObjectResolvingToVoid, as: "unaryObjectResolvingToVoid")
        connection.bind(binaryResolvingToIntCallback, as: "binaryResolvingToIntCallback")
        connection.bind(binaryResolvingToObjectCallback, as: "binaryResolvingToObjectCallback")
        connection.bind(binaryResolvingToObjectCallbackToInt, as: "binaryResolvingToObjectCallbackToInt")
    }

    struct JSAPITestStruct: Codable, Equatable {
        let intField = 42
        let stringField = "JSAPITEST"
    }
}

File: platforms/android/bridge-webview/src/androidTest/java/com/salesforce/nimbus/bridge/webview/MochaTests.kt
//
// Copyright (c) 2019, Salesforce.com, inc.
// All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
// For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
//

package com.salesforce.nimbus.bridge.webview

import android.util.Log
import android.webkit.WebView
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.internal.runner.junit4.statement.UiThreadStatement.runOnUiThread
import androidx.test.rule.ActivityTestRule
import com.salesforce.nimbus.BoundMethod
import com.salesforce.nimbus.JSONEncodable
import com.salesforce.nimbus.Plugin
import com.salesforce.nimbus.PluginOptions
import com.salesforce.nimbus.toJSONEncodable
import kotlinx.serialization.Serializable
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@RunWith(AndroidJUnit4::class)
class MochaTests {

    data class MochaMessage(val stringField: String = "This is a string", val intField: Int = 42) : JSONEncodable {
        override fun encode(): String {
            val jsonObject = JSONObject()
            jsonObject.put("stringField", stringField)
            jsonObject.put("intField", intField)
            return jsonObject.toString()
        }
    }

    @Serializable
    data class SerializableMochaMessage(val stringField: String = "This is a string", val intField: Int = 42)

    @PluginOptions(name = "mochaTestBridge")
    class MochaTestBridge(private val webView: WebView) : Plugin {

        val readyLatch = CountDownLatch(1)
        val completionLatch = CountDownLatch(1)

        // Set to -1 initially to indicate we never got a completion callback
        var failures = -1

        @BoundMethod
        fun ready() {
            readyLatch.countDown()
        }

        @BoundMethod
        fun testsCompleted(failures: Int) {
            this.failures = failures
            completionLatch.countDown()
        }
        @BoundMethod
        fun onTestFail(testTitle: String, errMessage: String) {
            Log.e("MOCHA", "[$testTitle]: $errMessage")
        }

        @BoundMethod
        fun sendMessage(name: String, includeParam: Boolean) {
            webView.post {
                var arg: JSONEncodable? = null
                if (includeParam) {
                    arg = MochaMessage()
                }
                webView.broadcastMessage(name, arg)
            }
        }
    }

    @Rule
    @JvmField
    val activityRule: ActivityTestRule<WebViewActivity> = ActivityTestRule<WebViewActivity>(
        WebViewActivity::class.java, false, true)

    @Test
    fun runMochaTests() {

        val webView = activityRule.activity.webView
        val testBridge =
            MochaTestBridge(
                webView
            )
        val jsAPITest = JSAPITestPlugin()

        val bridge = WebViewBridge()

        runOnUiThread {
            bridge.add(CallbackTestPluginWebViewBinder(CallbackTestPlugin()))
            bridge.add(MochaTestBridgeWebViewBinder(testBridge))
            bridge.add(JSAPITestPluginWebViewBinder(jsAPITest))
            bridge.attach(webView)
            webView.loadUrl("file:///android_asset/test-www/index.html")
        }

        assertTrue(testBridge.readyLatch.await(30, TimeUnit.SECONDS))

        runOnUiThread {
            webView.evaluateJavascript("""
            const titleFor = x => x.parent ? (titleFor(x.parent) + " " + x.title) : x.title
            mocha.run(failures => { __nimbus.plugins.mochaTestBridge.testsCompleted(failures); })
                 .on('fail', (test, err) => __nimbus.plugins.mochaTestBridge.onTestFail(titleFor(test), err.message));
            true;
            """.trimIndent()) {}
        }

        assertTrue(testBridge.completionLatch.await(30, TimeUnit.SECONDS))

        assertEquals(0, testBridge.failures)
    }

    @Test
    fun testExecutePromiseResolved() {
        val webView = activityRule.activity.webView
        val testBridge = MochaTestBridge(webView)

        val bridge = WebViewBridge()
        val callbackTestBinder = CallbackTestPluginWebViewBinder(CallbackTestPlugin())

        runOnUiThread {
            bridge.add(callbackTestBinder)
            bridge.add(MochaTestBridgeWebViewBinder(testBridge))
            bridge.attach(webView)
            webView.loadUrl("file:///android_asset/test-www/index.html")
        }

        assertTrue(testBridge.readyLatch.await(5, TimeUnit.SECONDS))
        val completionLatch = CountDownLatch(1)
        runOnUiThread {
            bridge.invoke(
                "__nimbus.plugins.callbackTestPlugin.addOne",
                args = arrayOf(5.toJSONEncodable())
            ) { err, result ->
                assertNull(err)
                assertEquals(6, result)
                completionLatch.countDown()
            }
        }

        assertTrue(completionLatch.await(5, TimeUnit.SECONDS))
    }

    @Test
    fun testExecutePromiseRejected() {
        val webView = activityRule.activity.webView
        val testBridge = MochaTestBridge(webView)

        val bridge = WebViewBridge()
        val callbackTestBinder = CallbackTestPluginWebViewBinder(CallbackTestPlugin())

        runOnUiThread {
            bridge.add(MochaTestBridgeWebViewBinder(testBridge))
            bridge.add(callbackTestBinder)
            bridge.attach(webView)
            webView.loadUrl("file:///android_asset/test-www/index.html")
        }

        assertTrue(testBridge.readyLatch.await(5, TimeUnit.SECONDS))
        val completionLatch = CountDownLatch(1)
        runOnUiThread {
            bridge.invoke(
                "__nimbus.plugins.callbackTestPlugin.failWith",
                arrayOf("epic fail".toJSONEncodable())
            ) { err, result ->
                assertEquals("epic fail", err)
                assertNull(result)
                completionLatch.countDown()
            }
        }

        assertTrue(completionLatch.await(5, TimeUnit.SECONDS))
    }

    @Test
    fun testPromiseRejectedOnRefresh() {
        val webView = activityRule.activity.webView
        val testBridge = MochaTestBridge(webView)

        val bridge = WebViewBridge()
        val callbackTestBinder = CallbackTestPluginWebViewBinder(CallbackTestPlugin())

        runOnUiThread {
            bridge.add(MochaTestBridgeWebViewBinder(testBridge))
            bridge.add(callbackTestBinder)
            bridge.attach(webView)
            webView.loadUrl("file:///android_asset/test-www/index.html")
        }

        assertTrue(testBridge.readyLatch.await(5, TimeUnit.SECONDS))
        val completionLatch = CountDownLatch(1)
        runOnUiThread {
            bridge.invoke(
                "__nimbus.plugins.callbackTestPlugin.wait",
                arrayOf(60000.toJSONEncodable())
            ) { err, _ ->
                assertEquals("ERROR_PAGE_UNLOADED", err)
                completionLatch.countDown()
            }
        }

        runOnUiThread {
            // Destroy the existing web page & JS context
            webView.loadUrl("file:///android_asset/test-www/index.html")
        }

        assertTrue(completionLatch.await(5, TimeUnit.SECONDS))
    }
}

data class JSAPITestStruct(var stringField: String = "JSAPITEST", var intField: Int = 42) : JSONEncodable {
    override fun encode(): String {
        val jsonObject = JSONObject()
        jsonObject.put("stringField", stringField)
        jsonObject.put("intField", intField)
        return jsonObject.toString()
    }

    companion object {
        fun decode(jsonString: String): JSAPITestStruct {
            val json = JSONObject(jsonString)
            val intField = json.getInt("intField")
            val stringField = json.getString("stringField")
            return JSAPITestStruct(stringField, intField)
        }
    }
}

@PluginOptions(name = "jsapiTestPlugin")
class JSAPITestPlugin : Plugin {

    @BoundMethod
    fun nullaryResolvingToInt(): Int {
        return 5
    }

    @BoundMethod
    fun nullaryResolvingToIntArray(): ArrayList<Int> {
        return arrayListOf(1, 2, 3)
    }

    @BoundMethod
    fun nullaryResolvingToObject(): JSAPITestStruct {
        return JSAPITestStruct()
    }

    @BoundMethod
    fun unaryResolvingToVoid(param: Int) {
        assertEquals(param, 5)
    }

    @BoundMethod
    fun unaryObjectResolvingToVoid(param: JSAPITestStruct) {
        assertEquals(param, JSAPITestStruct())
    }

    @BoundMethod
    fun binaryResolvingToIntCallback(param0: Int, param1: (result: Int) -> Unit) {
        assertEquals(param0, 5)
        param1(5)
    }

    @BoundMethod
    fun binaryResolvingToObjectCallback(param0: Int, param1: (result: JSAPITestStruct) -> Unit) {
        assertEquals(param0, 5)
        param1(JSAPITestStruct())
    }
}
